pkg/accesscontrol/access_store_test.go

```diff
@@ -274,6 +276,81 @@ func TestAccessStore_AccessFor(t *testing.T) {
 	}
 }
 
+type spyCache struct {
+	accessStoreCache
+
+	mu       sync.Mutex
+	setCalls map[any]int
+}
+
+func (c *spyCache) Add(k interface{}, v interface{}, ttl time.Duration) {
+	defer c.observeAdd(k)
+
+	time.Sleep(1 * time.Millisecond) // allow other routines to wake up, simulating heavy load
+	c.accessStoreCache.Add(k, v, ttl)
+}
+
+func (c *spyCache) observeAdd(k interface{}) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.setCalls == nil {
+		c.setCalls = make(map[any]int)
+	}
+	c.setCalls[k]++
+}
+
+func TestAccessStore_AccessFor_concurrent(t *testing.T) {
+	t.Skipf("TODO - Add a fix for this test")
+	testUser := &user.DefaultInfo{Name: "test-user"}
+	asCache := &spyCache{accessStoreCache: cache.NewLRUExpireCache(100)}
+	store := &AccessStore{
+		roles: roleRevisionsMock(func(ns, name string) string {
+			return fmt.Sprintf("%s%srev", ns, name)
+		}),
+		usersPolicyRules: &policyRulesMock{
+			getRBFunc: func(s string) []*rbacv1.RoleBinding {
+				return []*rbacv1.RoleBinding{
+					makeRB("testns", "testrb", testUser.Name, "testrole"),
+				}
+			},
+			getFunc: func(_ string) *AccessSet {
+				return &AccessSet{
+					set: map[key]resourceAccessSet{
+						{"get", corev1.Resource("ConfigMap")}: map[Access]bool{
+							{Namespace: All, ResourceName: All}: true,
+						},
+					},
+				}
+			},
+		},
+		cache: asCache,
+	}
+
+	const n = 5 // observation showed cases with up to 5 (or more) concurrent queries for the same user
```

Comment:
Can you elaborate on what it is we think is the wrong behavior / concern?

I can see that `l.cache.Add` is being called multiple times, but it will always have the same ID so the underlying cache (`*LRUCache`) will only have one of those (for the same key).

So is the issue that we're computing multiple times the result sets for nothing?

Another issue I'm seeing with the current code is there's a race between computing the cacheKey and creating the resulting `AccessSet`. The cacheKey is created with set `A` of RBAC resources, and there might be some changes to RBAC resources that that would lead to the `AccessSet` being created with a set `B` of RBAC resources. Is this what this test is after?

Reply:
You are correct in both observations.
1. This test aims to tackle the first one, with the additional concern of not only the multiple computation for the same key, but also how it's used afterwards (e.g. `AccessSets` seem to be embedded into the `Schemas` type). There seems to be some sort of indirection that makes the original problem very hard to troubleshoot.
2. Indeed this is a separate problem I also notice but it does not have a straightforward solution. I could take a look after this and try to find a viable alternative. In any case, I believe the solution to problem 1) will mitigate the impact of this other problem to some extent.

Reply:
Following up on problem 2):
https://github.com/rancher/steve/pull/292